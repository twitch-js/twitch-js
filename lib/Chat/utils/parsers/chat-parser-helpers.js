"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spread = (this && this.__spread) || function () {
    for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
    return ar;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.username = exports.mods = exports.emoteSets = exports.emotes = exports.badges = exports.followersOnly = exports.broadcasterLanguage = exports.userType = exports.generalTimestamp = exports.generalBoolean = exports.generalNumber = exports.generalString = void 0;
var camelCase_1 = __importDefault(require("lodash/camelCase"));
var isFinite_1 = __importDefault(require("lodash/isFinite"));
var replace_1 = __importDefault(require("lodash/replace"));
var toLower_1 = __importDefault(require("lodash/toLower"));
var twitch_1 = require("../../../twitch");
exports.generalString = function (maybeMessage) {
    return typeof maybeMessage === 'string'
        ? replace_1.default(maybeMessage, /\\[sn]/g, ' ')
        : undefined;
};
exports.generalNumber = function (maybeNumber) {
    var number = parseInt(maybeNumber, 10);
    return isFinite_1.default(number) ? number : undefined;
};
exports.generalBoolean = function (maybeBoolean) { return maybeBoolean === '1'; };
exports.generalTimestamp = function (maybeTimestamp) {
    var timestamp = new Date(parseInt(maybeTimestamp, 10));
    return timestamp.toString() !== 'Invalid Date' ? timestamp : new Date();
};
exports.userType = function (maybeUserType) {
    return typeof maybeUserType === 'string' ? maybeUserType : undefined;
};
exports.broadcasterLanguage = function (maybeLanguage) {
    return typeof maybeLanguage === 'string' ? maybeLanguage : undefined;
};
exports.followersOnly = function (maybeFollowersOnly) {
    var followersOnlyAsNumber = parseInt(maybeFollowersOnly, 10);
    if (followersOnlyAsNumber === 0) {
        return true;
    }
    else if (followersOnlyAsNumber > 0) {
        return followersOnlyAsNumber;
    }
    return false;
};
/**
 * Badges tag
 * @see https://dev.twitch.tv/docs/irc/tags/#privmsg-twitch-tags
 */
exports.badges = function (maybeBadges) {
    return typeof maybeBadges === 'string'
        ? maybeBadges.split(',').reduce(function (parsed, badge) {
            var _a, _b, _c;
            var _d = __read(badge.split('/'), 2), rawKey = _d[0], value = _d[1];
            if (typeof value === 'undefined') {
                return parsed;
            }
            var key = camelCase_1.default(rawKey);
            if (key in twitch_1.BooleanBadges) {
                return __assign(__assign({}, parsed), (_a = {}, _a[key] = exports.generalBoolean(value), _a));
            }
            if (key in twitch_1.NumberBadges) {
                return __assign(__assign({}, parsed), (_b = {}, _b[key] = parseInt(value, 10), _b));
            }
            return __assign(__assign({}, parsed), (_c = {}, _c[key] = value, _c));
        }, {})
        : {};
};
/**
 * Emote tag
 * @see https://dev.twitch.tv/docs/irc/tags/#privmsg-twitch-tags
 */
exports.emotes = function (maybeEmotes) {
    if (typeof maybeEmotes !== 'string') {
        return [];
    }
    return maybeEmotes.split('/').reduce(function (emoteTag, emoteIndices) {
        var _a = __read(emoteIndices.split(':'), 2), id = _a[0], indices = _a[1];
        if (!id) {
            return emoteTag;
        }
        return __spread(emoteTag, indices.split(',').map(function (index) {
            var _a = __read(index.split('-'), 2), start = _a[0], end = _a[1];
            return { id: id, start: parseInt(start, 10), end: parseInt(end, 10) };
        }));
    }, []);
};
exports.emoteSets = function (maybeEmoteSets) {
    return typeof maybeEmoteSets === 'string' ? maybeEmoteSets.split(',') : [];
};
exports.mods = function (message) {
    var _a = __read(message.split(': '), 2), modList = _a[1];
    return modList.split(', ');
};
exports.username = function () {
    var maybeUsernames = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        maybeUsernames[_i] = arguments[_i];
    }
    return maybeUsernames.reduce(function (maybeUsername, name) {
        if (typeof name !== 'string') {
            return maybeUsername;
        }
        if (name === 'tmi.twitch.tv') {
            return 'tmi.twitch.tv';
        }
        return toLower_1.default(name).split('.')[0];
    }, undefined);
};
//# sourceMappingURL=chat-parser-helpers.js.map