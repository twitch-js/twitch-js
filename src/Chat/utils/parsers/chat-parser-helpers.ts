import camelCase from 'lodash/camelCase'
import isFinite from 'lodash/isFinite'
import replace from 'lodash/replace'
import toLower from 'lodash/toLower'

import { Badges, BooleanBadges, NumberBadges, EmoteTag } from '../../../twitch'

export const generalString = (maybeMessage: string) => {
  return typeof maybeMessage === 'string'
    ? replace(maybeMessage, /\\[sn]/g, ' ')
    : undefined
}

export const generalNumber = (maybeNumber: string) => {
  const number = parseInt(maybeNumber, 10)
  return isFinite(number) ? number : undefined
}

export const generalBoolean = (maybeBoolean: string) => maybeBoolean === '1'

export const generalTimestamp = (maybeTimestamp: string) => {
  const timestamp = new Date(parseInt(maybeTimestamp, 10))
  return timestamp.toString() !== 'Invalid Date' ? timestamp : new Date()
}

export const userType = (maybeUserType: string) => {
  return typeof maybeUserType === 'string' ? maybeUserType : undefined
}

export const broadcasterLanguage = (maybeLanguage: string) => {
  return typeof maybeLanguage === 'string' ? maybeLanguage : undefined
}

export const followersOnly = (maybeFollowersOnly: string) => {
  const followersOnlyAsNumber = parseInt(maybeFollowersOnly, 10)

  if (followersOnlyAsNumber === 0) {
    return true
  } else if (followersOnlyAsNumber > 0) {
    return followersOnlyAsNumber
  }

  return false
}

/**
 * Badges tag
 * @see https://dev.twitch.tv/docs/irc/tags/#privmsg-twitch-tags
 */
export const badges = (maybeBadges: string): Partial<Badges> => {
  return typeof maybeBadges === 'string'
    ? maybeBadges.split(',').reduce((parsed, badge) => {
        const [rawKey, value] = badge.split('/')

        if (typeof value === 'undefined') {
          return parsed
        }

        const key = camelCase(rawKey)

        if (key in BooleanBadges) {
          return { ...parsed, [key]: generalBoolean(value) }
        }

        if (key in NumberBadges) {
          return { ...parsed, [key]: parseInt(value, 10) }
        }

        return { ...parsed, [key]: value }
      }, {} as Partial<Badges>)
    : {}
}

/**
 * Emote tag
 * @see https://dev.twitch.tv/docs/irc/tags/#privmsg-twitch-tags
 */
export const emotes = (maybeEmotes: string) => {
  if (typeof maybeEmotes !== 'string') {
    return []
  }

  return maybeEmotes.split('/').reduce((emoteTag, emoteIndices) => {
    const [id, indices] = emoteIndices.split(':')

    if (!id) {
      return emoteTag
    }

    return [
      ...emoteTag,
      ...indices.split(',').map((index) => {
        const [start, end] = index.split('-')
        return { id, start: parseInt(start, 10), end: parseInt(end, 10) }
      }),
    ]
  }, [] as EmoteTag[])
}

export const emoteSets = (maybeEmoteSets: string) => {
  return typeof maybeEmoteSets === 'string' ? maybeEmoteSets.split(',') : []
}

export const mods = (message: string) => {
  const [, modList] = message.split(': ')
  return modList.split(', ')
}

export const username = (...maybeUsernames: any[]): string =>
  maybeUsernames.reduce((maybeUsername, name) => {
    if (typeof name !== 'string') {
      return maybeUsername
    }

    if (name === 'tmi.twitch.tv') {
      return 'tmi.twitch.tv'
    }

    return toLower(name).split('.')[0]
  }, undefined)
