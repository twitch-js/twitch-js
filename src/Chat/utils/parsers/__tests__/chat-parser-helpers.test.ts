import * as helpers from '../chat-parser-helpers'

describe('Chat/utils/parsers/helpers', () => {
  describe('usernameFromPrefix', () => {
    test('should return undefined by default', () => {
      const actual = helpers.username(true, 1)

      expect(actual).toBeUndefined()
    })
  })

  describe('broadcasterLanguage', () => {
    test('should return string', () => {
      const language = 'en'

      const actual = helpers.broadcasterLanguage(language)
      const expected = language

      expect(actual).toEqual(expected)
    })

    test('should return undefined when given non-strings', () => {
      const language = {}

      const actual = helpers.broadcasterLanguage(language)

      expect(actual).toBeUndefined()
    })
  })

  describe('followersOnly', () => {
    test('should return false by default', () => {
      const actual = helpers.followersOnly()
      const expected = false

      expect(actual).toBe(expected)
    })

    test('should return true if followers only is active', () => {
      const actual = helpers.followersOnly('0')
      const expected = true

      expect(actual).toBe(expected)
    })

    test('should return minimum followers time', () => {
      const actual = helpers.followersOnly('7')
      const expected = 7

      expect(actual).toBe(expected)
    })
  })

  describe('badges', () => {
    test('should not parse unknown badges', () => {
      const badge = 'unknownBadge'
      const version = 'unknownBadgeVersion'

      const actual = helpers.badges(`${badge}/${version}`)
      const expected = { [badge]: version }

      expect(actual).toEqual(expected)
    })

    test('should parse known badges', () => {
      const badges =
        'admin/1,broadcaster/1,globalMod/1,moderator/1,partner/1,premium/1,staff/1,subGifter/1,turbo/1,vip/1,bits/100,bitsLeader/3,subscriber/9'

      const actual = helpers.badges(badges)

      expect(actual).toMatchSnapshot()
    })
  })
})
