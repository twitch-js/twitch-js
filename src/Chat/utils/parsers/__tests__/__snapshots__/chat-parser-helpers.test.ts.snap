// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chat/utils/parsers/helpers badges should parse known badges 1`] = `
Object {
  "admin": true,
  "bits": 100,
  "bitsLeader": 3,
  "broadcaster": true,
  "globalMod": true,
  "moderator": true,
  "partner": true,
  "premium": true,
  "staff": true,
  "subGifter": true,
  "subscriber": 9,
  "turbo": true,
  "vip": true,
}
`;
